packages/dd-trace/src/appsec/downstream_requests.js

```diff
@@ -0,0 +1,252 @@
+let perRequestDownstreamAnalysisCount
+
+function enable (_config) {
+  config = _config
+  globalRequestCounter = 0n
+  perRequestBodyAnalysisCount = new WeakMap()
+  perRequestDownstreamAnalysisCount = new WeakMap()
+}
+
+function disable () {
+  config = null
+  globalRequestCounter = 0n
+  perRequestBodyAnalysisCount = new WeakMap()
+  perRequestDownstreamAnalysisCount = new WeakMap()
+}
+
+/**
+ * Determines whether the current downstream request/responses bodies should be sampled for analysis.
+ * @param {import('http').IncomingMessage} req inbound request associated with the downstream call.
+ * @returns {boolean} true when the downstream response body should be captured.
+ */
+function shouldSampleBody (req) {
+  globalRequestCounter = (globalRequestCounter + 1n) & UINT64_MAX
+
+  const {
+    maxDownstreamRequestBodyAnalysis,
+    downstreamRequestBodyAnalysisSampleRate
+  } = config.appsec.apiSecurity
+
+  const currentCount = perRequestBodyAnalysisCount.get(req) || 0
+  if (currentCount >= maxDownstreamRequestBodyAnalysis) {
+    return false
+  }
+
+  const samplingRate = Math.min(Math.max(downstreamRequestBodyAnalysisSampleRate, 0), 1)
+
+  const hashed = (globalRequestCounter * KNUTH_FACTOR) % UINT64_MAX
+  const threshold = BigInt(Math.round(samplingRate * Number(UINT64_MAX)))
+
+  return hashed <= threshold
+}
+
+/**
+ * Increments the number of downstream body analyses performed for the given request.
+ * @param {import('http').IncomingMessage} req inbound request associated with the downstream call.
+ */
+function incrementBodyAnalysisCount (req) {
+  const currentCount = perRequestBodyAnalysisCount.get(req) || 0
+  perRequestBodyAnalysisCount.set(req, currentCount + 1)
+}
+
+/**
+ * Extracts request data from the context for WAF analysis
+ * @param {object} ctx context for the outgoing downstream request.
+ * @returns {object} a map of addresses and request data.
+ */
+function extractRequestData (ctx) {
+  const addresses = {}
+
+  const options = ctx?.args?.options || {}
+
+  addresses[HTTP_OUTGOING_METHOD] = determineMethod(options.method)
+
+  const headers = options?.headers
+  if (headers && Object.keys(headers).length > 0) {
+    addresses[HTTP_OUTGOING_HEADERS] = headers
+  }
+
+  return addresses
+}
+
+/**
+ * Extracts response data for WAF analysis.
+ * @param {import('http').IncomingMessage} res downstream response object.
+ * @param {boolean} includeBody flag describing if the response body should be collected.
+ * @param {Buffer|string|object|null} responseBody response body.
+ * @returns {object} a map of addresses and response data.
+ */
+function extractResponseData (res, includeBody, responseBody) {
+  const addresses = {}
+
+  if (res.statusCode !== undefined && res.statusCode !== null) {
+    addresses[HTTP_OUTGOING_RESPONSE_STATUS] = String(res.statusCode)
+  }
+
+  const headers = res.headers
+  if (headers && Object.keys(headers).length > 0) {
+    addresses[HTTP_OUTGOING_RESPONSE_HEADERS] = headers
+  }
```

Comment:
Would it make sense to capture trailer headers (`res.trailers`) here so WAF can analyze the full response header set, including end-of-stream metadata?